### PR TITLE
[Build] Add CMake changes from conda-forge build

### PR DESCRIPTION
**Pull request by hmacdope**
## Description

Related to #5718

## Checklist
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [X] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [X] All changes have test coverage
- [X] Code is well-documented
- [X] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [X] Related issue is referred in this PR

## Changes
* Re-works CMake to allow building from external dependencies if required. 
* changes libxsmm to header only mode, seeking feedback on this part, we can add some more conditional complexity here if preferred.

### CODE REVIEW

**Comment by hmacdope** [comment]
@jermainewang @frozenbugs, this is the PR to upstream the conda-forge cmake changes, let me know what you think and I will incorporate as soon as I can.

**Comment by hmacdope** [comment]
Also tagging @mikemhenry

**Comment by frozenbugs** [comment]
@dgl-bot

**Comment by hmacdope** [comment]
```
libdgl.so: undefined reference to `dgemv_'
libdgl.so: undefined reference to `dgemm_'
libdgl.so: undefined reference to `sgemm_'
libdgl.so: undefined reference to `sgemv_'
```

Seems to be the issue in 3/75 of the builds. Are these symbols provided by complied libxsmm perhaps? Something else?

**Comment by frozenbugs** [comment]
@dgl-bot

**Comment by hmacdope** [comment]
Possibly related to  using / not using `-D_BLAS=0` https://libxsmm.readthedocs.io/en/latest/#link-instructions

**Comment by frozenbugs** [comment]
Hi @hmacdope, are you still working on resolving this issue? Let us know if you need any help from us. Thanks!

**Comment by hmacdope** [comment]
Yes please, I can't figure out exactly why some builds are missing the symbols. Any ideas?

**Comment by anko-intel** [comment]
> Possibly related to using / not using `-D_BLAS=0` https://libxsmm.readthedocs.io/en/latest/#link-instructions

You are right, according to https://github.com/libxsmm/libxsmm#header-only you have to define __BLAS=0
so, adding it to lines [217-221](https://github.com/dmlc/dgl/pull/6189/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR217):
```
if(USE_LIBXSMM)
  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -DUSE_LIBXSMM -DDGL_CPU_LLC_SIZE=40000000 -D__BLAS=0")
  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DUSE_LIBXSMM -DDGL_CPU_LLC_SIZE=40000000 -D__BLAS=0")
  message(STATUS "Build with LIBXSMM optimization.")
endif(USE_LIBXSMM)
```
seems to be enough

**Comment by mikemhenry** [comment]
@hmacdope can you make that change when you get a chance? I don't have write permission on your fork

**Comment by hmacdope** [comment]
@mikemhenry done. 

@frozenbugs should be good to re-trigger CI :)

**Comment by hmacdope** [comment]
Thanks so much for the fix @anko-intel, much appreciated. :)

**Comment by frozenbugs** [comment]
@dgl-bot

**Comment by frozenbugs** [comment]
@dgl-bot

File: CMakeLists.txt [hunk 1]
```diff
@@ -108,6 +117,19 @@ dgl_feature_option(
...
+if (EXTERNAL_DLPACK_PATH OR EXTERNAL_DMLC_PATH OR EXTERNAL_NANOFLANN_PATH OR EXTERNAL_NANOFLANN_PATH OR EXTERNAL_METIS_PATH OR EXTERNAL_GKLIB_PATH)
+  message(STATUS "Using at least one external library")
+  set(USE_EXTERNAL_LIBS ON)
+  
+  if (BUILD_CPP_TEST)
+    message(FATAL_ERROR "Cannot build cpp unittests with external libraries")
+  endif(BUILD_CPP_TEST)
+
+endif()
+
```
Comment (frozenbugs) [inline CMakeLists.txt:129]:
```suggestion
endif()
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -34,6 +34,15 @@ dgl_option(
     python3
 )
 
+dgl_option(EXTERNAL_DLPACK_PATH "Path to external dlpack" OFF)
```
Comment (frozenbugs) [inline CMakeLists.txt:37]:
```suggestion
# Conda build related options.
dgl_option(EXTERNAL_DLPACK_PATH "Path to external dlpack" OFF)
```

**Review by frozenbugs — APPROVED** [review]

File: cmake/modules/FindMETIS.cmake [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+FIND_LIBRARY(METIS_LIBRARIES
+    NAMES
+    libmetis metis
+    PATHS
+    ${EXTERNAL_METIS_LIB_PATH}
+    )
+
+
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(METIS DEFAULT_MSG METIS_INCLUDE_DIR METIS_LIBRARIES)
+MARK_AS_ADVANCED(METIS_LIBRARIES METIS_INCLUDE_DIR)
```
Comment (frozenbugs) [inline cmake/modules/FindMETIS.cmake:27]:
```suggestion
MARK_AS_ADVANCED(METIS_LIBRARIES METIS_INCLUDE_DIR)

```

**Comment by mikemhenry** [comment]
:tada: Awesome! I am really excited for the next dgl release

**Comment by hmacdope** [comment]
Thanks everyone!

**Pull request merged**